pkg/ruler/base/ruler.go

```diff
@@ -508,6 +551,9 @@ func (r *Ruler) listRules(ctx context.Context) (result map[string]rulespb.RuleGr
 	case r.cfg.ShardingStrategy == util.ShardingStrategyDefault:
 		result, err = r.listRulesShardingDefault(ctx)
 
+	case r.cfg.ShardingStrategy == util.ShardingStrategyByRule:
```

Comment:
Can you expand on your idea behind wanting shuffle sharding please? AFAICS shuffle-sharding won't help much here because operators (such as ourselves) typically run small numbers of `ruler`s, so I don't see much value in isolating rule evaluation for certain tenants to subsets of the ruler fleet. If we truly want tenant isolation, we should implement https://github.com/grafana/loki/pull/8129 as well and move the shuffle-sharding to the read path where it is more impactful.

Reply:
> Sharding by rule instead of group in loki seems strictly superior. Let's just change the all ownership checks to consider rule instead of group -- WDYT? That way everyone gets the benefit of improved sharding and we don't have to expose another option.

I like this idea, but for later - maybe v3.0 once this feature has had time in production and we're confident in it.

Reply:
> Can you expand on your idea behind wanting shuffle sharding please? AFAICS shuffle-sharding won't help much here because operators (such as ourselves) typically run small numbers of rulers, so I don't see much value in isolating rule evaluation for certain tenants to subsets of the ruler fleet. If we truly want tenant isolation, we should implement https://github.com/grafana/loki/pull/8129 as well and move the shuffle-sharding to the read path where it is more impactful.

Actually I care less about shuffle sharding per se for rules. My only concern is given we already have shuffle sharding option, this new "sharding by rule" should respect that and should co-exist together IMO. People who are already using shuffle sharding with multi-tenants they cannot benefit from this "shard by rule".

Also `shard by rule` seems to be separate concern to me independent of tenants or users. It should work for everyone independent of multi-tenants or shuffle sharding.

Reply:
I think that's a fair point.

However, I suspect that the shuffle-sharding option is not being used much, and even if it is then https://github.com/grafana/loki/pull/8129 will achieve the same thing really by pushing the shuffle-sharding where it really belongs (IMHO) on the read path, and those users can then enable `by-rule` sharding if they need it. WDYT?

Reply:
Yea that sounds reasonable. Understand no point of considering "shuffle sharding" and complicate things once #8129 is implemented.

I'm convinced in that case given we communicate our intention clearly in changelog and release notes :)

Reply:
I've implemented this request in https://github.com/grafana/loki/pull/8092/commits/88b55deea959b4355cbf340b1dc51d61f652fe1e, PTAL